virtcontainers/fc.go

```diff
@@ -546,6 +555,77 @@ func (fc *firecracker) fcSetVMBaseConfig(mem int64, vcpus int64, htEnabled bool)
 	return err
 }
 
+func (fc *firecracker) fcSetLogger() error {
+	span, _ := fc.trace("fcSetLogger")
+	defer span.Finish()
+
+	fcLogLevel := "Error"
+
+	var err error
+	fcLogFifoFile := filepath.Join(fc.vmPath, fcLogFifo)
+	fc.logFifo, err = fifo.OpenFifo(context.Background(), fcLogFifoFile, syscall.O_CREAT|syscall.O_RDONLY|syscall.O_NONBLOCK, 0)
+	if err != nil {
+		return fmt.Errorf("Failed to open/create firecracker log fifo file")
+	}
+
+	jailedLogFifo, err := fc.fcJailResource(fcLogFifoFile, fcLogFifo)
+	if err != nil {
+		fc.Logger().WithField("set logger failed", err).Error()
+		return err
+	}
+
+	// listen on log fifo file and transfer error info
+	go func() {
+		scanner := bufio.NewScanner(fc.logFifo)
+		for scanner.Scan() {
+			fc.Logger().WithFields(logrus.Fields{
+				"level":    "log",
+				"contents": scanner.Text()}).Error("firecracker failed")
+		}
+
+		if err := scanner.Err(); err != nil {
+			fc.Logger().WithError(err).Errorf("Failed reading firecracker logs from fifo file")
+		}
+	}()
+
+	fcMetricsFifoFile := filepath.Join(fc.vmPath, fcMetricsFifo)
+	fc.metricsFifo, err = fifo.OpenFifo(context.Background(), fcMetricsFifoFile, syscall.O_CREAT|syscall.O_RDONLY|syscall.O_NONBLOCK, 0)
+	if err != nil {
+		return fmt.Errorf("Failed to open/create firecracker metrics fifo file")
+	}
+
+	jailedMetricsFifo, err := fc.fcJailResource(fcMetricsFifoFile, fcMetricsFifo)
+	if err != nil {
+		fc.Logger().WithField("set logger failed", err).Error()
+		return err
+	}
+
+	// listen on metrics file and transfer error info
+	go func() {
+		scanner := bufio.NewScanner(fc.metricsFifo)
+		for scanner.Scan() {
+			fc.Logger().WithFields(logrus.Fields{
+				"level":    "metrics",
+				"contents": scanner.Text()}).Error("firecracker failed")
```

Comment:
Same question as above. Why are metrics logged as errors as well.

Reply:
yep, metrics data will be more about performance, but it also has data structure like `drive_fails`,  `machine_cfg_fails`, ... etc for users to spot root cause of failure. And Like I said last comment, I have set the level as `Error`, it will only output when firecracker fail.

Reply:
thanks @Pennyzct for the explanation.